### PR TITLE
- Fixes zonemaster/zonemaster-engine#504

### PR DESCRIPTION
**Pull request by vlevigneron**
- Inhibits for the moment blacklisting mechanism. We need to fix zonemaster/zonemaster-engine#438 before considering to re-activate it.
- It could slower execution time of very broken zones.
- The code is still there, we just have added a flag to disable it (re-activitaion will be trivial).

### CODE REVIEW

**Comment by matsduf** [comment]
@vlevigneron! The problem occurs when the server sends a REFUSED and we take that as no response. There is one Test Case that could trigger REFUSED and that is NAMESERVER01 because we query for a name completely outside the zone in question. It is normal for a authoritative name server to respond with REFUSED in that case.

I all other cases we query for names in or below the zone, and then the name server will not respond with REFUSED unless it is not configured with the zone, and then it will consistently respond with REFUSED and blacklisting should not be a problem.

Is it possible to turn blacklisting off for just NAMESERVER01?

Blacklisting is more complicated than above, there is potentially unwanted blacklisting unrelated to the problem of #438 and that is elaborated in #513.

**Comment by vlevigneron** [comment]
@matsduf I agree with you in the fact that reply with a REFUSED can be a legitimate answer from a server. The problem here is that this is something that is never tested in all test cases we have (we never explicitely check REFUSED rcode returned). And this is exactly what happens, we query for '.', then the server answer with a REFUSED that we interpret as a no response , as a consequence we blacklist this server.

That's why it is more complicated than just this case that I propose to remove, and came back later when we will have fixed LDNS part or find a workaround that allow us to make a difference between nos response and REFUSED.

I can add a flag to allow the code to disable blacklisting on demand instead of a global switch. Is it what you propose for this patch ?

**Comment by vlevigneron** [comment]
Here is my proposal (it will be pushed in minutes) : 
- I kept the global switch (in case we decide to completely remove blacklisting mechanism).
- I added an on demand flag that should be used for queries that are supposed to answer with legitimate REFUSED (like Nameserver07). That will prevent to add the server in the black list. That flag should be used for all queries of names completely outside the zone we are testing.

File: lib/Zonemaster/Engine/Constants.pm [hunk 1]
```diff
@@ -67,6 +68,8 @@ Readonly our $ALGO_STATUS_OTHER        => 5;
 Readonly our $ALGO_STATUS_DELETE_DS    => 6;
 Readonly our $ALGO_STATUS_INDIRECT_KEY => 7;
 
+Readonly our $BLACKLISTING_ENABLED     => 1;
```
Comment (mattias-p) [inline lib/Zonemaster/Engine/Constants.pm]: I don't think it makes much sense to have this as Readonly. Please consider either making it writable or removing it.

File: lib/Zonemaster/Engine/Nameserver.pm [hunk 2]
```diff
@@ -393,9 +393,11 @@ sub _query {
             $msg =~ s/$trailing_info.*/\./;
             Zonemaster::Engine->logger->add( LOOKUP_ERROR =>
                   { message => $msg, ns => "$self", name => "$name", type => $type, class => $href->{class} } );
-            $self->blacklisted->{ $flags{usevc} }{ $flags{dnssec} } = 1;
-            if ( !$flags{dnssec} ) {
-                $self->blacklisted->{ $flags{usevc} }{ !$flags{dnssec} } = 1;
+            if ( not $href->{q{blacklisting_disabled}} ) {
```
Comment (mattias-p) [inline lib/Zonemaster/Engine/Nameserver.pm]: Could you add some documentation for this new option?

Comment (vlevigneron) [reply lib/Zonemaster/Engine/Nameserver.pm]: OK. Just after lunch.

File: lib/Zonemaster/Engine/Constants.pm (hunk 1 above)
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Constants.pm]: It makes sense for me. It's a global switch that can not be modified. I remind you this is a temporary hack and the whole mechanism should be reviewed and modified accordingly when we will have find out whats wrong in zonemaster/LDNS interaction for "no responses"/REFUSED cases. This flag should be removed at this occasion.

File: lib/Zonemaster/Engine/Nameserver.pm (hunk 2 above)
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Nameserver.pm]: Is it enough for you ?

Comment (mattias-p) [reply lib/Zonemaster/Engine/Nameserver.pm]: Yeah, looks good :)

**Comment by matsduf** [comment]
Nameserver01 must also have the "blacklist no" setting, and in both cases (Nameserver01 and Nameserver07) this is not only for now, but permanently.

**Comment by vlevigneron** [comment]
@matsduf just did it for Nameserver01

**Pull request merged**